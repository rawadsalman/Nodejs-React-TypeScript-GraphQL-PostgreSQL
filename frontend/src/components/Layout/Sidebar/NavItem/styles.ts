import { keyframes } from '@emotion/react'
import styled from '@emotion/styled'

const transition = '350ms ease'
export const NAVITEM_HEIGHT = 64
export const SIDEBAR_WIDTH = 78

export const flyIn = keyframes`
  0% { transform: rotate(0deg); left: -200px}
  80% { transform: rotate(0deg); left: ${SIDEBAR_WIDTH}px; }
  100%{ transform: rotate(5deg); left: ${SIDEBAR_WIDTH}px; }
`

const NavItemInnerDiv = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  background: #1e272e;
  z-index: 2;
  transition: background-color ${transition};

  svg {
    fill: white;
    path {
      fill: white;
      transition: fill ${transition};
    }
  }

  &.active {
    svg {
      path {
        fill: ${({ theme }) => theme.colors.primary};
      }
    }
  }

  &:hover {
    background-color: #2a343a;
  }
`

const NavItemOuterDiv = styled.div`
  position: relative;
  display: block;
  height: ${NAVITEM_HEIGHT}px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;

  &:hover {
    cursor: pointer;

    .flyout {
      animation: ${flyIn} 250ms ease 1;

      left: ${SIDEBAR_WIDTH}px;
      transform: rotate(2deg);
      opacity: 1;
    }
  }
`

const NavItemFlyout = styled.div`
  z-index: 1;
  position: absolute;
  height: ${NAVITEM_HEIGHT}px;
  min-width: ${SIDEBAR_WIDTH}px;
  padding: 0 16px;
  overflow: hidden;

  opacity: 1;
  left: -500px;
  /* transition: all 250ms ease; */

  animation: none;

  font-family: Raleway;
  font-size: 1.1rem;
  color: white;

  background: transparent;
  text-transform: uppercase;
  border-top-right-radius: 8px;
  border-bottom-right-radius: 8px;
  text-shadow: -1px -1px 0 black, 1px -1px 0 black, -1px 1px 0 black, 1px 1px 0 black,
    1px 0px 0px ${({ theme }) => theme.colors.primary}, 0px 1px 0px ${({ theme }) => theme.colors.primary},
    2px 1px 0px ${({ theme }) => theme.colors.primary}, 1px 2px 0px ${({ theme }) => theme.colors.primary},
    3px 2px 0px ${({ theme }) => theme.colors.primary}, 2px 3px 0px ${({ theme }) => theme.colors.primary},
    4px 3px 0px ${({ theme }) => theme.colors.primary}, 3px 4px 0px ${({ theme }) => theme.colors.primary},
    5px 4px 0px ${({ theme }) => theme.colors.primary}, 3px 5px 0px black, 6px 5px 0px black, -1px 2px 0 black,
    0 3px 0 black, 1px 4px 0 black, 2px 5px 0px black, 2px -1px 0 black, 3px 0 0 black, 4px 1px 0 black,
    5px 2px 0px black, 6px 3px 0 black, 7px 4px 0 black;

  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
`

export { NavItemInnerDiv, NavItemOuterDiv, NavItemFlyout }
