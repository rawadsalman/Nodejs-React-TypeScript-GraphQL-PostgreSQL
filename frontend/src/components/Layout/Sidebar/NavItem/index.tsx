import React from 'react'
import { useHistory, useRouteMatch } from 'react-router-dom'
import { NavItemFlyout, NavItemInnerDiv, NavItemOuterDiv } from './styles'

interface INavItem {
  to: string
  activeOnlyWhenExact?: boolean
  displayName?: string
}

const NavItem: React.FC<INavItem> = ({ children, to, activeOnlyWhenExact, displayName }) => {
  let match = useRouteMatch({
    path: to,
    exact: activeOnlyWhenExact,
  })
  let history = useHistory()

  const handleClick = () => history.push(to)

  return (
    <NavItemOuterDiv className={`${match ? 'active' : ''}`} onClick={handleClick}>
      <NavItemInnerDiv className={`${match ? 'active' : ''}`}>{children}</NavItemInnerDiv>
      <NavItemFlyout className="flyout">
        <span>{displayName ? displayName : to.substr(1)}</span>
      </NavItemFlyout>
    </NavItemOuterDiv>
  )
}

export default NavItem
