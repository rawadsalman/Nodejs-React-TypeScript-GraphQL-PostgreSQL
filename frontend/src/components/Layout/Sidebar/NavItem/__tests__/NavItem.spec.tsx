import { render, fireEvent } from '@testing-library/react'
import React from 'react'
import NavItem from '../index'
import { ThemeProvider } from '@emotion/react'

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
  useRouteMatch: () => ({}),
}))

describe('NavItem', () => {
  it('matches Snapshot', async () => {
    const wrapper = render(
      <ThemeProvider theme={{ colors: { primary: 'red' } }}>
        <NavItem to="yolo">test</NavItem>
      </ThemeProvider>
    )

    const { getByText } = wrapper

    expect(wrapper).toMatchSnapshot()
    fireEvent.click(getByText('test'))
  })
})
