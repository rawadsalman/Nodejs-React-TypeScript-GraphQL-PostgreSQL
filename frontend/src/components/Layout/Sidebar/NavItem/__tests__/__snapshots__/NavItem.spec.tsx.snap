// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="active css-1imfco6-NavItemOuterDiv ek60gb91"
      >
        <div
          class="active css-1slwpb5-NavItemInnerDiv ek60gb92"
        >
          test
        </div>
        <div
          class="flyout css-ur38sj-NavItemFlyout ek60gb90"
        >
          <span>
            olo
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="active css-1imfco6-NavItemOuterDiv ek60gb91"
    >
      <div
        class="active css-1slwpb5-NavItemInnerDiv ek60gb92"
      >
        test
      </div>
      <div
        class="flyout css-ur38sj-NavItemFlyout ek60gb90"
      >
        <span>
          olo
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
